sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -312,6 +312,16 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+  val REDUCE_POST_SHUFFLE_PARTITIONS_FOR_REPARTITION =
```

Comment:
I probably won't add this config. If users call `Dataset#repartition`, we have to respect it and not change it.

Reply:
I initially didn't add this config.  Later I found with that change, the `test("Union two datasets with different pre-shuffle partition number")` won't test what we want to test, because it was written using repartition. Probably I can rewrite that test without using repartition.

Reply:
let's rewrite the tests. We can use `spark.range` to create 2 relations with different num partitions

Reply:
I think we can remove that test. Excluding repartition, the pre-shuffle num partitions are always the same (200 by default).

We do need to have a test for repartition though.